instrumentation/r2dbc-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/r2dbc/v1_0/TraceProxyListener.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.r2dbc.v1_0;
+
+import static io.opentelemetry.javaagent.bootstrap.Java8BytecodeBridge.currentContext;
+import static io.opentelemetry.javaagent.instrumentation.r2dbc.v1_0.R2dbcSingletons.instrumenter;
+
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.r2dbc.proxy.core.QueryExecutionInfo;
+import io.r2dbc.proxy.listener.ProxyMethodExecutionListener;
+import io.r2dbc.spi.ConnectionFactoryOptions;
+import java.util.HashMap;
+import java.util.Map;
+
+public class TraceProxyListener implements ProxyMethodExecutionListener {
+
+  private final ConnectionFactoryOptions factoryOptions;
+  private final Map<QueryExecutionInfo, DbExecution> dbExecutions = new HashMap<>();
+
+  public TraceProxyListener(ConnectionFactoryOptions factoryOptions) {
+    this.factoryOptions = factoryOptions;
+  }
+
+  @Override
+  public void beforeQuery(QueryExecutionInfo queryInfo) {
+    Context parentContext = currentContext();
+    DbExecution dbExecution = new DbExecution(queryInfo, factoryOptions);
+    dbExecutions.put(queryInfo, dbExecution);
+    if (!instrumenter().shouldStart(parentContext, dbExecution)) {
+      return;
+    }
+    Context context = instrumenter().start(parentContext, dbExecution);
+    dbExecution.setContext(context);
+    Scope scope = parentContext.makeCurrent();
```

Comment:
Opening scope makes sense only if you can guarantee that the `afterQuery` method is called on the same thread. It is useful when the span that you are creating here had a child span. I think I would not open a scope here.